feat: add new backend `BitbucketDataCenterOAuth2`

eshaan7 opened the pull request:
## Proposed changes

Resolves https://github.com/python-social-auth/social-core/issues/855.
Related social-docs PR: https://github.com/python-social-auth/social-docs/pull/190

- Adds a new backend `BitbucketDataCenterOAuth2`.
- Generalizes the PKCE logic in a base class `BaseOAuth2PKCE` and re-used in both `TwitterOAuth2` and `BitbucketDataCenterOAuth2`.
- Generalizes the PKCE unit-tests logic in base classes `OAuth2PkcePlainTest` and `OAuth2PkceS256Test` and re-used in both `test_twitter_oauth2.py` and `test_bitbucket_datacenter.py`

## Types of changes

Please check the type of change your PR introduces:

- [ ] Release (new release request)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (PEP8, lint, formatting, renaming, etc)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Build related changes (build process, tests runner, etc)
- [ ] Other (please describe):

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask. We're here to
help! This is simply a reminder of what we are going to look for before merging
your code._

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added documentation to https://github.com/python-social-auth/social-docs

## Other information

Any other information that is important to this PR such as screenshots of how
the component looks before and after the change.

nijel commented:
Abstracting that sounds like a good idea. Please also add tests and documentation.

eshaan7 commented:
@nijel; 

Can I get a review on this now? I have added complete tests along with refactoring the existing PKCE ones. Just docs pending, will do it later.

nijel commented on social_core/tests/backends/test_twitter_oauth2.py:
```diff
@@ -172,58 +170,23 @@ def test_login(self):
         self.strategy.set_settings(
             {
-                "SOCIAL_AUTH_TWITTER_OAUTH2_PKCE_CODE_CHALLENGE_METHOD": "invalidmethodname"
+                "SOCIAL_AUTH_"
+                + self.name
+                + "_PKCE_CODE_CHALLENGE_METHOD": "invalidmethodname"
```
This might be better to read:

```suggestion
                f"SOCIAL_AUTH_{self.name}_PKCE_CODE_CHALLENGE_METHOD": "invalidmethodname"
```

nijel reviewed (approved):
Thanks, looks good now. I think we can merge this once the docs is ready as well.

eshaan7 commented:
@nijel; pushed some smaller changes addressing the review. Made `USE_PKCE` a setting for Bitbucket Data Center since it supports both with and without.

Also, docs PR is up at: https://github.com/python-social-auth/social-docs/pull/190

johanv26 commented on social_core/backends/oauth.py:
```diff
@@ -459,3 +462,69 @@ def refresh_token(self, token, *args, **kwargs):
+        return code_verifier
+
+    def get_code_verifier(self):
+        name = self.name + "_code_verifier"
```
```suggestion
        name = f"{self.name}_code_verifier"
```

johanv26 commented on social_core/backends/oauth.py:
```diff
@@ -459,3 +462,69 @@ def refresh_token(self, token, *args, **kwargs):
+    DEFAULT_USE_PKCE = True
+
+    def create_code_verifier(self):
+        name = self.name + "_code_verifier"
```
```suggestion
        name = f"{self.name}_code_verifier"
```

johanv26 commented on social_core/backends/oauth.py:
```diff
@@ -459,3 +462,69 @@ def refresh_token(self, token, *args, **kwargs):
+            encoded = base64.urlsafe_b64encode(hashed)
+            code_challenge = encoded.decode().replace("=", "")  # remove padding
+            return code_challenge
+        elif method == "plain":
+            return code_verifier
+        else:
+            raise AuthException("Unsupported code challenge method.")
```
```suggestion
        if method == "plain":
            return code_verifier
        raise AuthException("Unsupported code challenge method.")
```

johanv26 reviewed (commented):
Add some style suggestions

eshaan7 commented:
@nijel; Let's merge this :D

nijel commented:
Merged, thanks for your contribution!

eshaan7 commented:
@nijel; Thank you! Can I expect this to be released to pypi anytime soon?
